extensions/vertx-http/deployment/src/test/resources/conf/http-roles-grant-config.properties

```diff
@@ -0,0 +1,27 @@
+quarkus.http.auth.basic=true
+quarkus.http.auth.policy.t1.roles.admin=Admin1
+quarkus.http.auth.permission.t1.paths=/*
+quarkus.http.auth.permission.t1.policy=t1
+quarkus.http.auth.policy.t2.roles.admin=Admin2
+quarkus.http.auth.permission.t2.paths=/*
+quarkus.http.auth.permission.t2.policy=t2
+quarkus.http.auth.policy.t3.roles.user=Janet,Robin
+quarkus.http.auth.policy.t3.roles.admin=Janet,Robin
+quarkus.http.auth.permission.t3.paths=/test/multiple-new-roles-1
+quarkus.http.auth.permission.t3.policy=t3
+quarkus.http.auth.policy.t4.roles.user=Monica,Robin
+quarkus.http.auth.policy.t4.roles.admin=Monica,Robin
+quarkus.http.auth.permission.t4.paths=/test/multiple-new-roles-2,/test/multiple-new-roles-3
+quarkus.http.auth.permission.t4.policy=t4
+quarkus.http.auth.policy.t5.roles-allowed=admin
+quarkus.http.auth.policy.t5.roles.admin=Admin1
+quarkus.http.auth.permission.t5.paths=/test/roles-allowed-path
+quarkus.http.auth.permission.t5.policy=t5
+quarkus.http.auth.policy.t6.roles-allowed=Admin1
+quarkus.http.auth.policy.t6.roles.admin=Admin1
+quarkus.http.auth.permission.t6.paths=/test/always-denied
```

Comment:
`roles-allowed` allows an access so I'm not why it is tested against `/test-always-denied` ?

Reply:
I've been trying to formulate a question, I think I have it now, so what happens when we have
```
quarkus.http.auth.policy.t5.roles-allowed=Admin1
quarkus.http.auth.policy.t5.roles.admin=Admin1
quarkus.http.auth.permission.t5.paths=/test/roles-allowed-path
```

but no `@RolesAllowed` annotation ? It still should work if the token has an `admin` role, right ?

Reply:
It won't work because the `SecurityIdentity` with role `admin` does not have role `Admin1` at the moment when policy is applied.

- There is exactly one path matching policy applied on every request, let's call it policy ONE.
- Policy ONE only grants roles if it is successful (that is: if you have required roles; ** == authenticated)
- If you want, I can turn it around and first add roles.
- If I make it work, what will be result? Following policy:

```
quarkus.http.auth.policy.t6.roles-allowed=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.permission.t6.paths=/test/always-denied
```

does exactly same thing as this policy:

```
quarkus.http.auth.policy.t6.roles-allowed=admin
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.permission.t6.paths=/test/always-denied
```

Right now there should be clear message to users: if you have `quarkus.http.auth.policy.t6.roles-allowed=admin`, we will grant `quarkus.http.auth.policy.t6.roles` or `quarkus.http.auth.policy.t6.permissions`.

Maybe this is documentation issue? I thought it is clear enough now that I added note about `@RolesAllowed`.

Reply:
let me ask differently @sberyozkin , what is the advantage of granting roles first? this policy basically means that `admin` role implies `Admin1` role, so I don't know why should users check for `roles-allowed=Admin1`.

Reply:
for record, I can imagine something like this

```
quarkus.http.auth.policy.t6.roles-allowed=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.permission.t6.paths=/test/always-denied
```

Reply:
Hi @michalvavrik Sorry if I'm a bit slow, so, if I have
```
quarkus.http.auth.policy.t5.roles-allowed=Admin1
quarkus.http.auth.policy.t5.roles.admin=Admin1
quarkus.http.auth.permission.t5.paths=/test/roles-allowed-path
```
then why can't the authorization filter for this policy check that there is a mapping between `SecurityIdentity` `admin` and `Admin1` is available ? You say:

>It won't work because the SecurityIdentity with role admin does not have role Admin1 at the moment when policy is applied.

But it works if you have `@RolesAllowed` - I'm assuming that both `quarkus.http.auth.policy.t5.roles-allowed=Admin1` and `@RolesAllowed("Admin1")`  - users can choose either of these 2 options, are evaluated at the same time ?

This example, 
```
quarkus.http.auth.policy.t6.roles-allowed=admin
quarkus.http.auth.policy.t6.roles.admin=Admin1
```

does not really work or rather, `quarkus.http.auth.policy.t6.roles.admin=Admin1` has no any meaning here since `Admin1` is not used.

In the end all I'm asking for is to ensure that whether users choose to use `@RolesAllowed("MappedRole")` or `quarkus.http.auth.policy.t6.roles-allowed=MappedRole`, where there is a mapping `roles.mapped-role=MappedRole`, it works in both cases, for consistency purposes, even if using the mapping option without `@RolesAllowed` seems not very practical.

Reply:
> Hi @michalvavrik Sorry if I'm a bit slow, so, if I have

nah, I just don't know how to explain this. Important thing is that if you don't understand it, users won't as well.

> then why can't the authorization filter for this policy check that there is a mapping between SecurityIdentity admin and Admin1 is available ?

My thought is that policy `t5` adds role `Admin1` only if it is complied. Especially because of the A / B thingy I explain in the next paragraph.

> `quarkus.http.auth.policy.t6.roles-allowed=admin \n quarkus.http.auth.policy.t6.roles.admin=Admin1`does not really work or rather, quarkus.http.auth.policy.t6.roles.admin=Admin1 has no any meaning here since Admin1 is not used.

I disagree. If A implicated B, that you can use A instead of B for check that would only apply on B. And I say: you can use `admin` instead of `Admin1`, because at the point you are granting role `Admin1`, you already know it has role `admin`, so why do you want to check it with `roles-allowed`.

Why it differs for `@RolesAllowed`? Because in the policy, you already know the identity has these roles, but when it comes to the `@RolesAllowed` check, you are starting from a scratch.

> In the end all I'm asking for is to ensure that whether users choose to use @RolesAllowed("MappedRole") or quarkus.http.auth.policy.t6.roles-allowed=MappedRole, where there is a mapping roles.mapped-role=MappedRole, it works in both cases, for consistency purposes, even if using the mapping option without @RolesAllowed seems not very practical.

I'll make it work. Please let's conclude this thread, it's not a big deal at all.

Reply:
Truth be told, you can be right, because there might be policy set like this:

```
quarkus.http.auth.policy.t6.roles-allowed=Admin1
quarkus.http.auth.policy.t6.roles.admin=Admin1
quarkus.http.auth.policy.t6.roles.root=role1
quarkus.http.auth.permission.t6.paths=/test/*
```

and the `SecurityIdentity` with role `root` may have `Admin1`, but not role `admin`.

Reply:
Done.